api/src/opentrons/protocol_runner/legacy_command_mapper.py

```diff
@@ -191,78 +228,231 @@ def _build_initial_command(
+                wellName=well_name,
+            ),
+        )
+
+    def _build_pick_up_tip_command(
+        self,
+        command: legacy_command_types.PickUpTipMessage,
+        command_id: str,
+        now: datetime,
+    ) -> pe_commands.Command:
+        pipette: LegacyPipetteContext = command["payload"]["instrument"]
+        location = command["payload"]["location"]
+        if isinstance(location, LegacyWell):
+            well = location
+        elif location.labware.is_well:
+            well = location.labware.as_well()
+        mount = MountType(pipette.mount)
+        #   the following type checking suppression assumes the tiprack is not loaded on top of a module
+        slot = DeckSlotName.from_primitive(well.parent.parent)  # type: ignore[arg-type]
+        well_name = well.well_name
+        labware_id = self._labware_id_by_slot[slot]
+        pipette_id = self._pipette_id_by_mount[mount]
+
+        return pe_commands.PickUpTip.construct(
+            id=command_id,
+            key=command_id,
+            status=pe_commands.CommandStatus.RUNNING,
+            createdAt=now,
+            startedAt=now,
+            params=pe_commands.PickUpTipParams.construct(
+                pipetteId=pipette_id,
+                labwareId=labware_id,
+                wellName=well_name,
+            ),
+        )
+
+    def _build_liquid_handling_commands(
+        self,
+        command: Union[
+            legacy_command_types.AspirateMessage, legacy_command_types.DispenseMessage
+        ],
+        command_id: str,
+        now: datetime,
+    ) -> pe_commands.Command:
+        pipette: LegacyPipetteContext = command["payload"]["instrument"]
+        location = command["payload"]["location"]
+        volume = command["payload"]["volume"]
+        flow_rate = command["payload"]["rate"]
+        if isinstance(location, Location) and location.labware.is_empty is False:
+            if location.labware.is_well:
                 well = location.labware.as_well()
-            else:
-                raise Exception("Unknown drop_tip location.")
+            slot = DeckSlotName(location.labware.first_parent())
+            parent_module_id = self._module_id_by_slot.get(slot)
+            labware_id = (
+                self._labware_id_by_module_id[parent_module_id]
+                if parent_module_id is not None
+                else self._labware_id_by_slot[slot]
+            )
             mount = MountType(pipette.mount)
-            slot = DeckSlotName.from_primitive(well.parent.parent)  # type: ignore[arg-type]
             well_name = well.well_name
-            labware_id = self._labware_id_by_slot[slot]
             pipette_id = self._pipette_id_by_mount[mount]
-            engine_command = pe_commands.DropTip.construct(
+
+            if command["name"] == legacy_command_types.ASPIRATE:
+                if flow_rate == 1.0:
+                    flow_rate = command["payload"]["rate"] * pipette.flow_rate.aspirate
```

Comment:
Why is there an `if flow_rate == 1.0:` here? This multiplication still needs happen if the payload is e.g. `0.9` or `1.1`

```suggestion
                flow_rate = command["payload"]["rate"] * pipette.flow_rate.aspirate
```

Reply:
Yeah, I had the same comment. Moving it here:

> Why is this calculation under an if-statement? I'm pretty sure we need to do this calculation no matter the value of `flow_rate`: 1.0, 2.0, 0.5, or anything else.
> 
> If `pipette.flow_rate.aspirate` is 123, and the protocol does `pipette.aspirate(volume, rate=2.0)`, I'm pretty sure the absolute flow rate will be 246, so the Protocol Engine `aspirate` command should be constructed with `flowRate=246`. Does that match your understanding?

Reply:
Hey yall, thanks! I see Mike's comment below about how `130` isn't a valid `rate`. I thought you could input a `flow_rate` up to the max volume of the pipette for some reason. But maybe this is false knowledge that I always used since it probably caps at the real max `rate`. This makes a lot of sense.